### PR TITLE
Optimize related model queries

### PR DESCRIPTION
**Pull request by horiaradu**
Optimise related model queries

No longer query for the related/included model when no results are obtained in the first query

fix: https://github.com/strongloop/loopback-datasource-juggler/issues/1006

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by Amir-61** [comment]
@slnode test please

File: lib/include.js [hunk 1]
```diff
@@ -220,7 +220,11 @@ Inclusion.include = function(objects, include, options, cb) {
       pageSize = inqLimit;
     }
     if (pageSize <= 0) {
-      return model.find(filter, options, cb);
+      if (size) {
+        return model.find(filter, options, cb);
+      } else {
+        cb(null, []);
+      }
```
Comment (Amir-61) [inline lib/include.js]: How about simplyifing this to by early returning:

```
 if (size) {
   return model.find(filter, options, cb);
 } 
 cb(null, []);
```

File: lib/include.js [hunk 2]
```diff
@@ -248,11 +252,15 @@ Inclusion.include = function(objects, include, options, cb) {
...
+      if (foreignKeys.length) {
+        model.find(newFilter, options, function(err, results) {
+          if (err) return done(err);
+          items = items.concat(results);
+          done();
+        });
+      } else {
         done();
-      });
+      }
```
Comment (Amir-61) [inline lib/include.js]: How about this:
```
if (!foreignKeys.length) {
  return done();
}
model.find(newFilter, options, function(err, results) {
  if (err) return done(err);
  items = items.concat(results);
  done();
});
```

File: lib/include.js [hunk 3]
```diff
@@ -626,6 +634,7 @@ Inclusion.include = function(objects, include, options, cb) {
         defineCachedRelations(obj);
         obj.__cachedRelations[relationName] = [];
       }
+
```
Comment (Amir-61) [inline lib/include.js]: Please don't change unrelated code

File: lib/include.js [hunk 4]
```diff
@@ -876,6 +886,7 @@ Inclusion.include = function(objects, include, options, cb) {
         defineCachedRelations(obj);
         obj.__cachedRelations[relationName] = null;
       }
+
```
Comment (Amir-61) [inline lib/include.js]: ditto

**Comment by Amir-61** [comment]
There is a commit linter issue:
```
**  2 problems found:
**    c6bae99 - Optimize related model queries: line 3 longer than 72 characters.
**    c6bae99 - Optimize related model queries: line 5 longer than 72 characters.
```
@horiaradu Please fix it.

**Comment by Amir-61** [comment]
The changes look mostly good to me. @bajtos any thought?

**Comment by horiaradu** [comment]
@Amir-61 if I run the tests & lint locally, everything checks ok and I can't access the linter error https://ci.sl.strongloop.com/job/strongloop-commit-linter-jobs/job/loopback-datasource-juggler/20//console.

File: lib/include.js [hunk 5]
```diff
@@ -248,6 +251,9 @@ Inclusion.include = function(objects, include, options, cb) {
       newFilter.where[fkName] = {
         inq: foreignKeys,
       };
+      if (!foreignKeys.length) {
```
Comment (bajtos) [inline lib/include.js]: Can we move this short-circuit logic higher, so that we can skip creation of `newFilter` object altogether? I would personally apply this filter on `listOfFKs` before calling `async.each`.

```js

// Optimization: no need to resolve keys that are an empty array
listOfFKs = listOfFKs.filter(function(keys) { return keys.length > 0; });

async.each(listOfFKs, ...);
```

File: lib/include.js [hunk 6]
```diff
@@ -220,7 +220,10 @@ Inclusion.include = function(objects, include, options, cb) {
       pageSize = inqLimit;
     }
     if (pageSize <= 0) {
-      return model.find(filter, options, cb);
+      if (size) {
```
Comment (bajtos) [inline lib/include.js]: Similarly here, can we check & return earlier, e.g. on L214?

```js
if (!foreignKeys.length) {
  return cb(null, []);
}

if (filter.limit || ...);
...
```

**Comment by bajtos** [comment]
Looks mostly good to me, I left two comments on how to improve the code a bit.

Looks like our CI is failing on amazon/Node@v0.12, could you PTAL @Amir-61?

**Comment by Amir-61** [comment]
>@Amir-61 if I run the tests & lint locally, everything checks ok and I can't access the linter error https://ci.sl.strongloop.com/job/strongloop-commit-linter-jobs/job/loopback-datasource-juggler/20//console.

Not sure why you cannot see the link; @rmg I have seen many people from the community say they cannot see link `strongloop-commit-linter-jobs`, any thought?

@horiaradu the linter is complaining that the comit description is longer than 72 chars; please break them to smaller sentences

Right now this is your commit:
```
Optimize related model queries 
  
No longer query for the related/included model when no results are obtained in the first query
```

You can change it to:
```
Optimize related model queries 
  
No longer query for the related/included model
when no results are obtained in the first query
```

**Comment by Amir-61** [comment]
@slnode test please

**Comment by Amir-61** [comment]
Now it needs your PR to be rebased as well:
```
PR out of date, needs to be rebased
```

File: lib/include.js [hunk 7]
```diff
@@ -234,6 +237,9 @@ Inclusion.include = function(objects, include, options, cb) {
     }
 
     var items = [];
+    listOfFKs = listOfFKs.filter(function(keys) {
```
Comment (Amir-61) [inline lib/include.js]: Please add the comment suggested by bajtos to make it more clear
```
// Optimization: no need to resolve keys that are an empty array
listOfFKs = listOfFKs.filter(function(keys) { return keys.length > 0; });
```

**Comment by Amir-61** [comment]
@slnode test please

**Comment by rmg** [comment]
I've adjusted the permissions. @horiaradu are you able to see the [Commit Linter](https://ci.sl.strongloop.com/job/strongloop-commit-linter-jobs/job/loopback-datasource-juggler/98/console) and [PR Linter](https://ci.sl.strongloop.com/job/strongloop-pr-linter-jobs/job/loopback-datasource-juggler/99/console) results now?

**Comment by Amir-61** [comment]
@slnode test please

**Comment by Amir-61** [comment]
>I've adjusted the permissions. @horiaradu are you able to see the Commit Linter and PR Linter results now?

Thanks @rmg. I hear that many people from the community cannot see these files? Can we adjust it in a way that everyone can see them instead of giving permission for just specific people?

**Comment by Amir-61** [comment]
@slnode test please

**Comment by rmg** [comment]
> Can we adjust it in a way that everyone can see them instead of giving permission for just specific people?

That is what I did.

**Comment by horiaradu** [comment]
Still can't see the PR/commit linter

**Comment by rmg** [comment]
@horiaradu I've opened the permissions even further now. Can you try viewing [this PR linter log](https://ci.sl.strongloop.com/job/strongloop-pr-linter-jobs/job/loopback-datasource-juggler/104/console)? Note that you'll need to login using github.

**Comment by horiaradu** [comment]
it works now. thanks @rmg !

**Comment by Amir-61** [comment]
@slnode test please

**Comment by Amir-61** [comment]
Thanks @horiaradu for contribution.

@jannyHou and @superkhau 

ummm... it seems the code base for `2.x` is very different from `master` branch for this area; there is no function called `findWithForeignKeysByPage()` where this optimization is being applied in `2.x`; I' doubt if these optimization can be back-ported to `2.x`; thought?

CC: @bajtos
